From 031a5b97265bfc26a891269efaf94ab255e44664 Mon Sep 17 00:00:00 2001
From: Phil Staub <Phil.Staub@windriver.com>
Date: Fri, 19 Oct 2012 12:52:52 -0700
Subject: [PATCH] mips64: Add support for generic MIPS64R2 processor

QEMU for 64-bit MIPS provides a single MIPS64R2 processor selection:
MIPS64R2-generic, reporting Processor ID 0x00010000. This encoding
corresponds to the MIPS Technologies (MTI) code in the company field
with the Processor ID and Revision subfields left blank.

When cpu_probe() is invoked, it dispatches to cpu_probe_mips(), which
in turn has no handler for the blank Processor ID/Revision fields, so
no valid cputype is written in the cpuinfo_mips structure, and the
boot process cannot continue.

Here we add some definitions and a switch case to enable sane handling
of this processor (pseudo)type.

Finally, adding MIPS64R2-generic support to the kernel deserves to be
recognized in the Kconfig.

Signed-off-by: Phil Staub <Phil.Staub@windriver.com>
---
 arch/mips/Kconfig            |    1 +
 arch/mips/include/asm/cpu.h  |    3 ++-
 arch/mips/kernel/cpu-probe.c |    4 ++++
 3 files changed, 7 insertions(+), 1 deletion(-)

diff --git a/arch/mips/Kconfig b/arch/mips/Kconfig
index 8862b57..b9da3c75 100644
--- a/arch/mips/Kconfig
+++ b/arch/mips/Kconfig
@@ -283,6 +283,7 @@ config MIPS_MALTA
 	select SYS_HAS_CPU_MIPS32_R1
 	select SYS_HAS_CPU_MIPS32_R2
 	select SYS_HAS_CPU_MIPS64_R1
+	select SYS_HAS_CPU_MIPS64_R2
 	select SYS_HAS_CPU_NEVADA
 	select SYS_HAS_CPU_RM7000
 	select SYS_HAS_EARLY_PRINTK
diff --git a/arch/mips/include/asm/cpu.h b/arch/mips/include/asm/cpu.h
index f9fa2a4..5026a1e 100644
--- a/arch/mips/include/asm/cpu.h
+++ b/arch/mips/include/asm/cpu.h
@@ -79,6 +79,7 @@
  * These are the PRID's for when 23:16 == PRID_COMP_MIPS
  */
 
+#define PRID_IMP_MIPS64R2_GENERIC	0x0000	/* QEMU */
 #define PRID_IMP_4KC		0x8000
 #define PRID_IMP_5KC		0x8100
 #define PRID_IMP_20KC		0x8200
@@ -267,7 +268,7 @@ enum cpu_type_enum {
 	 */
 	CPU_5KC, CPU_20KC, CPU_25KF, CPU_SB1, CPU_SB1A, CPU_LOONGSON2,
 	CPU_CAVIUM_OCTEON, CPU_CAVIUM_OCTEON_PLUS, CPU_CAVIUM_OCTEON2,
-	CPU_XLR, CPU_XLP,
+	CPU_XLR, CPU_XLP, CPU_MIPS64R2_GENERIC,
 
 	CPU_LAST
 };
diff --git a/arch/mips/kernel/cpu-probe.c b/arch/mips/kernel/cpu-probe.c
index 5099201..193ec78 100644
--- a/arch/mips/kernel/cpu-probe.c
+++ b/arch/mips/kernel/cpu-probe.c
@@ -792,6 +792,10 @@ static inline void cpu_probe_mips(struct cpuinfo_mips *c, unsigned int cpu)
 {
 	decode_configs(c);
 	switch (c->processor_id & 0xff00) {
+	case PRID_IMP_MIPS64R2_GENERIC:
+		c->cputype = CPU_MIPS64R2_GENERIC;
+		__cpu_name[cpu] = "MIPS64R2-generic";
+		break;
 	case PRID_IMP_4KC:
 		c->cputype = CPU_4KC;
 		__cpu_name[cpu] = "MIPS 4Kc";
-- 
1.7.9.7

